packages/compose/src/hooks/use-drop-zone/index.js

```diff
@@ -219,20 +219,17 @@ export default function useDropZone( {
 			ownerDocument.addEventListener( 'dragenter', maybeDragStart );
 
 			return () => {
-				onDropRef.current = null;
-				onDragStartRef.current = null;
-				onDragEnterRef.current = null;
-				onDragLeaveRef.current = null;
-				onDragEndRef.current = null;
-				onDragOverRef.current = null;
```

Comment:
Gave this a quick look, and I can see the mem leak again:

https://user-images.githubusercontent.com/1083581/191437281-e5c883d3-48a0-41d8-b9ba-d8d3c4ce443b.mp4

> Happy to separate this into another PR if necessary.

Up to you — it may be beneficial to tackle this separately in case this PR had to be merged urgently

Reply:
Thanks for testing! Do you have the exact reproduction steps? I would like to give it a look!

This PR is not urgent so we can take it easy 😆.

Reply:
> Do you have the exact reproduction steps?

From the video just posted above, and my original report (https://github.com/WordPress/gutenberg/pull/38999#pullrequestreview-891717184), I think we could start with:

1. Run the editor on your local machine (DEV MODE), to make sure that React fires dev-only warnings
2. Create a new post
3. Add some text to the first paragraph
4. Drag an image and drop it in the editor, under the first paragraph
5. Select the paragraph block, click and drag its drag handle and drag the paragraph under the image block
6. React would have printed the warning to console about a state update within `DropZoneComponent`

Reply:
I actually think that this can be also reproduced on `trunk`, I saw a similar react warning when working on #44301

Reply:
The problem seems to be that an update in one of the listeners in `useDropZone` triggers a  remount of the `<MediaPlaceholder>` component. It then triggers `useRefEffect` to fire the cleanup function. Due to a fundamental flaw in the `useRefEffect` implementation, it remounts the component **before** firing the remaining events. This usually isn't a problem, but we register another event in one of those event handlers. This causes the added event handlers to be _stalled_ and never cancel throughout the component's lifecycle.

However, if we change the `useRefEffect` implementation to `useEffect` somehow, React will make sure to run the remaining event listener before calling the cleanup effect. I think maybe this is why it's traditionally discouraged to register event listeners outside `useEffect`?

Obviously, we can't change how `useRefEffect` work now as that would impact a lot of the existing code, and I'm not aware of a way to fix it properly either. I might have some idea for this particular bug though but it won't be a general solution 😅 .

Reply:
That does sound complicated 😅 

I agree that we should definitely not touch `useRefEffect` for now. 

Maybe we can try to work around the problem by re-architecting how ev listeners are added/removed in `MediaPlaceholder`, or whatever other approach you had in mind?

Also, this may have become big/complex enough to deserve its own separate PR?

Reply:
I tried to fix it in https://github.com/WordPress/gutenberg/pull/42722/commits/49387fc6b16a45d80ca9e8da02a02f70e566aa34 by not registering/removing event listeners in another listeners, the timing they are called is probably not what we thought it would.

In short, before that commit:
1. We register the `dragenter` event (on `document`) when the `ref` callback is attached to some element.
2. Once the `dragenter` event is fired, we immediately remove the event itself.
3. On `drop` or `dragend` event, we re-attach the `dragenter` event.

In https://github.com/WordPress/gutenberg/pull/42722/commits/49387fc6b16a45d80ca9e8da02a02f70e566aa34, we simply remove the last two steps and keep the listeners untouched. It would only get removed once the element is detached from DOM.

It works great in my testing, please LMK if it works for you 🙏! If you (or anyone else) think this is too complicated and deserve a separate PR, I'll be happy to do so too :).

Reply:
Wait, I was wrong, this can still happen in `useEffect` in some cases 🤔 . Either way the commit fixes the issue though.

Reply:
I guess the general rule of thumb is to not add event listeners inside another event listener. That could cause the listener to be stalled and never get cleaned up when the component unmounts.

Reply:
> It works great in my testing, please LMK if it works for you 🙏! If you (or anyone else) think this is too complicated and deserve a separate PR, I'll be happy to do so too :).

Tested on `gutenberg.run` (as [I'm currently unable to do so locally](https://github.com/WordPress/gutenberg/issues/44416)), it seems to work well!

Definitely much cleaner anyway — I would probably just the `isDragging` variable in order to know what to do inside each ev listener, instead of adding/removing listeners inside listeners.

Taking a look at the code, I also noticed that a few `ownerDocument.addEventListener` calls are also happening inside another ev listener — should we refactor those too?

Reply:
> should we refactor those too?

Yeah, probably! I'd do so in another PR and probably refactor `use-drop-zone` completely to hide this complexity. An idea that came to mind is to encapsulate most of the event handling in a class and just initiate that class in a `useEffect`. That way we can ensure the listeners are not out of our control when the component unmounts.

Reply:
Sounds good, curious to see how this approach feels like. Please keep me in the loop when you work on that!